src/interfaces/index.ts

```diff
@@ -1,5 +1,56 @@
 export interface IPaginationOptions {
+  /**
+   * the amount of items to be requested per page
+   */
   limit: number;
+  /**
+   * the page that is requested
+   */
   page: number;
+  /**
+   * a babasesic route for generating links (i.e., WITHOUT query params)
+   */
   route?: string;
 }
+
+export interface IPaginationMeta {
```

Comment:
Ok! So the main reason I delayed my review was because I was trying to find some documentation! (Plus mega busy!) 

Firstly! I like the idea and if this was a project where only you and I were using it I was say "Yes! This is great! I like the organised structure!" however being open source I feel we should stick to some 'standards' in terms of pagination responses. I've seen countless different methods of implementing pagination response objects so I wanted to hunt down the REST thesis and find a section that explains pagination. However there doesn't appear to be much mention on the internet other than some broken links 🤔.   
So! Having said that and failing at finding some 'best practice' guid or 'standards' it's probably best to stick to the most common, most agreed on structure. 

So! From searching the internet I've found that `links` key is commonly used, sometimes as `_links`. `metadata` is a bit hit and miss. I personally like it! However I don't think it's a common key found in pagination responses. If you have any documentation that says otherwise! Please share! I'd rather keep this! 

This argument also brings up the subject of key renaming and custom structures. Might be cool if we could add an optional method (callback) that can manipulate the structure and or keys of the response object. I think that is a BIG change though to some of the responses and possibly too much? And possibly an additional PR? What do you think?

Reply:
Dear @bashleigh , i am not sure if i understand your concerns with this. Basically you want me to remove the `IPaginationMeta` interface?

I see this package as a helper to receive some paginated results from the database - and nothing more! When it comes to outputting data, i would **not** use this package at all - i would rather transform the paginated entities. In this case, i would then add the `metadata` to my output structure.

I mainly added the `meta` key in order to not bloat the main object.

Because you asked for some documents / standards in this domain. Unfortunately, there is no common ground. REST itself is rather an architectural style (for building and application) and not a communication protocol (as for example SOAP was). However, people often misunderstand this and think, REST will give clear indications, how to format messages and so on. However, there are some "best practices" in certain areas - for example `{json:api}` is a rather well-known "standard" on how to structure request / response documents (see https://jsonapi.org ). In json:api you can also add `meta` to your response (i.e., see the official top-level-element here: https://jsonapi.org/format/#document-top-level and an example here https://jsonapi.org/examples/#pagination ).

Reply:
That's what I meant! I was hoping REST would implement some standards but it doesn't. No many people stick to the thesis anyway :joy: 

Cool! I'm happy with those supplied docs. In such case, could you change the name of `items` to `data` for consistency with these docs and add a section to the read me about it? I'd much rather build this package alongside some sort of standard rather than guessing (which I'm good at!).

But in the future I think we should defo implement a key name and structure changing callback. Because, let's say you install a frontend package and the required structure is different to that of this supplied package. You'll have to build your own transformer. I'd rather make it so more people can use this package and do what they want to achieve :)

Reply:
Dear @bashleigh ,
i am not sure with changing the name of `items` to `data`. I think `{json:api}` is just a way of formatting your request / response messages. In particular, it offers a few nice features - if you fully commit to this "standard".

However, i see `{json:api}` as a `view model` for outputting data - that is completely separated from the `internal structure` of your models.

Take, for example, the famous `fractal` php package (i know, php :D) here https://fractal.thephpleague.com/

This package allows for transforming your internal data structure to an output structure (i.e., modify content, omit fields, ...) and then serialize it to different output formats (like, for example, `json:api`. The flow would be as follows:

1) get your (paginated) data
2) shove it into fractal
3) output the structure of fractal

So basically, fractal just works with the data you shove into it - that is not coupled to the pagination module that returns the data from the storage.

What do you think?

Reply:
It's the closest thing to a 'standard' that we have by the looks if it :joy: yeyy! 

yes! I'm not talking about the structure of the model! I mean the structure of pagination! So I'm imagining something like 

```ts
return await paginate<MyCustomPagination<Entity>>(this.repo, {page, limit}, {}, (paginationObject: (paginationObject: Pagination<Entity>): MyCustomPagination<Entity> => ({ data: paginationObject.items, metadata: {...paginationObject.meta, somethingCustom: true,}}));
```

So you can customise and or change the structure of the pagination object :) not to transform the entities :p however! You mentioning that! This is what I have to do when using pagination with nestjsx/crud when transforming models. If I implemented the above I could keep the responses consistent! 

```ts
 Array.isArray(objects[objects.hasOwnProperty('data') ? 'data' : 'items'])
      ? {
          ...objects,
          [objects.hasOwnProperty('items') ? 'items' : 'data']: objects[
            objects.hasOwnProperty('items') ? 'items' : 'data'
          ].map((object: any) => this.handleOne(object)),
        }
      : this.handleOne(objects)
```